pkg/cached/cache_test.go

```diff
@@ -0,0 +1,833 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cached_test
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"sort"
+	"strings"
+	"testing"
+
+	"k8s.io/kube-openapi/pkg/cached"
+)
+
+func TestSource(t *testing.T) {
+	count := 0
+	source := cached.NewSource(func() cached.Result[[]byte] {
+		count += 1
+		return cached.NewResultOK([]byte("source"), "source")
+	})
+	if err := source.Get().Err; err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+	if err := source.Get().Err; err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+	if count != 2 {
+		t.Fatalf("Expected function called twice, called: %v", count)
+	}
+}
+
+func TestSourceError(t *testing.T) {
+	count := 0
+	source := cached.NewSource(func() cached.Result[[]byte] {
+		count += 1
+		return cached.NewResultErr[[]byte](errors.New("source error"))
+	})
+	if err := source.Get().Err; err == nil {
+		t.Fatalf("expected error, found none")
+	}
+	if err := source.Get().Err; err == nil {
+		t.Fatalf("expected error, found none")
+	}
+	if count != 2 {
+		t.Fatalf("Expected function called twice, called: %v", count)
+	}
+}
+
+func TestSourceAlternate(t *testing.T) {
+	count := 0
+	source := cached.NewSource(func() cached.Result[[]byte] {
+		count += 1
+		if count%2 == 0 {
+			return cached.NewResultErr[[]byte](errors.New("source error"))
+		} else {
+			return cached.NewResultOK([]byte("source"), "source")
+		}
+	})
+	if err := source.Get().Err; err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+	if err := source.Get().Err; err == nil {
+		t.Fatalf("expected error, found none")
+	}
+	if err := source.Get().Err; err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+	if err := source.Get().Err; err == nil {
+		t.Fatalf("expected error, found none")
+	}
+	if count != 4 {
+		t.Fatalf("Expected function called 4x, called: %v", count)
+	}
+}
+
+func TestStaticSource(t *testing.T) {
+	count := 0
+	source := cached.NewStaticSource(func() cached.Result[[]byte] {
+		count += 1
+		return cached.NewResultOK([]byte("source"), "source")
+	})
+	if err := source.Get().Err; err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+	if err := source.Get().Err; err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+	if count != 1 {
+		t.Fatalf("Expected function called once, called: %v", count)
+	}
+}
+
+func TestStaticSourceError(t *testing.T) {
+	count := 0
+	source := cached.NewStaticSource(func() cached.Result[[]byte] {
+		count += 1
+		return cached.NewResultErr[[]byte](errors.New("source error"))
+	})
+	if err := source.Get().Err; err == nil {
+		t.Fatalf("expected error, found none")
+	}
+	if err := source.Get().Err; err == nil {
+		t.Fatalf("expected error, found none")
+	}
+	if count != 1 {
```

Comment:
hm I thought it said errors were always re-called?

Reply:
Not the `NewStaticSource` no

Reply:
https://github.com/kubernetes/kube-openapi/pull/345/files#diff-a0198b5ee0e3e127b9a9b56a1fc535e84a0fc7e9fe9d789befed02641834bf00R216-R217